### ISSUE TITLE
Use a clipping library to clean up / improve handling of slice generated polys.

### ISSUE THREAD

**Issue by DanielJoyce** [Pacmanfan/UVDLPSlicerController#14]
Robust polygon clipping library, would be very useful in managing and cleaning up polys generated from slightly messy geometry or overlapping objects. Better than trying to write our own

http://sourceforge.net/projects/polyclipping/files/?source=navbar

zip includes c# version

I'll look into seeing how hard it would be to integrate, the existing clipper is pretty bare bones, which does help with integration, not a lot of things to change.

**Comment by DanielJoyce** [comment]
It does use fixed point math, so we'd have to translate floats into that format. But 3d printers can't print anything below a certain resolution anyways. Fixed point / integer math is numerically robust though. 

We will pay some performance penalty when translating to integer coordinates, but gain a lot in final polygon quality and handling edge cases.

**Comment by Pacmanfan** [comment]
That's a good find, although the slicer in Creation Workshop actually takes care of much of the bad and overlapping geometry of the models. What I would actually like to get working  better is my 3d CSG code. I can intersect, subtract, and union 3d meshes together with it.  I have some basic object generation code too such as cube, sphere, cylinder. Take a look in the /3dengine/csg directory

**Comment by Pacmanfan** [comment]
I'd be wary about trying to transform things to fixed-point. I wrote an entire 3d engine using fixed point years ago. How good are you with OpenGL? What I really need to do is speed up object translation / rotation /scaling/ etc... Currently, the objects don't use a matrix to hold these, I'm simply operating on the points - and having to regenerate the GL Call Lists every time something changes.

**Comment by DanielJoyce** [comment]
Yeah, thats way more difficult, trying to directly handle csg code. csgjs is not numerically stable, and even openscad regularly has problems. Unless there is a port of CGAL to C#, getting it even remotely right so it doesn't barf all the time is difficult. Even openscad and its javascript clone have models that cause them to choke all the time due to numerical stability issues.

**Comment by DanielJoyce** [comment]
Doesn't OpenTK provide routines to handle that? At slicing time, its usually simpler to bake the transformations in though. I'll dig into that bit too.

**Comment by Pacmanfan** [comment]
Yes, that's what I was concerned about, If I start moving / scaling /rotating objects using a matrix, then I would probably have to flatten it all out back into the points before scaling. The other concern was object hit-testing via mouse-click on screen and object intersection for support generation.

**Comment by DanielJoyce** [comment]
Those are solvable. Really, at slice time we'd back the transformations in on a clone of the mesh and start slicing.

**Comment by DanielJoyce** [comment]
Transforming into 'fixed point' / integer is easy. The end result of using a library like this is we can handle intersecting/simplifying slices of meshes with (self) intersecting volumes, and performing 'slice unions' that have the result of effectively 'unioning' the models.

**Comment by DanielJoyce** [comment]
So are the slicer notes inaccurate then?

https://github.com/Pacmanfan/UVDLPSlicerController/blob/master/UVDLP/Software/PC/UV_DLP_3dPrinter/UV_DLP_3D_Printer/Slicing/OldSlicer/OldSliceNotes.txt

Specificly

> The problem I am currently encounting is that sometimes, geometry overlaps
this causes areas that should be filled to not be filled and vice versa
The *optimal solution to this is to make sure geometries never overlap
we can do this in several ways either in 3d as a union operation
or in 2d as a crossings test somehow
we can also do this in 2d by rendering each and every object separately, then ORing the
resultant images together. this approach probably would be the simplest, but at an additonal 
time cost.

because the first question is due to the selection of winding rule, the second can be solved by union of slice polygons, and to have it be robust, we need to use a robust library like Clipper.

**Comment by Pacmanfan** [comment]
Yes, I will update the notes. The slicer has gone through a lot of improvement since Beta 13 over the past year. As far as I know, The slicer is working perfectly at the moment.

-Steve

**Comment by shaise** [comment]
Hi Steve, Daniel,

Maybe the slicer can merge with the new SVG slicing system.

Shai
